cpp/src/arrow/io/io-file-test.cc

```diff
@@ -462,6 +462,41 @@ TEST_F(TestMemoryMappedFile, ReadOnly) {
   rommap->Close();
 }
 
+TEST_F(TestMemoryMappedFile, ReadOver4GbFile) {
+    // ARROW-1096
+    const int64_t buffer_size = 1000 * 1000;
+    std::vector<uint8_t> buffer(buffer_size);
+
+    test::random_bytes(buffer_size, 0, buffer.data());
+
+    const int64_t reps = 5000;
+
+    std::string path = "ipc-read-over-4gb-file-test";
+    std::shared_ptr<MemoryMappedFile> rwmmap;
+    ASSERT_OK(InitMemoryMap(reps * buffer_size, path, &rwmmap));
+
+    int64_t position = 0;
+    for (int i = 0; i < reps; ++i) {
+        ASSERT_OK(rwmmap->Write(buffer.data(), buffer_size));
+        position += buffer_size;
+    }
+    rwmmap->Close();
```

Comment:
@wesm , it might be good to test that the whole range of allocated bytes in memory mapped file is writable/readable (since `mmap` implementation contains logic to calculate file offset and size independently). If that's fine I will change the unit test name from `ReadOver4GbFile` to `ReadWriteOver4GbFile`.